### PR TITLE
Fix HTML drop issues with the Windows browsers

### LINKED ISSUE #36824: DropZone component listeners not working on latest Chromium browser update
### Description

The onHTMLDrop event listener on the `DropZone` component seems to have stopped working on the latest version of Chrome (Version 96.0.4664.45 (Official Build) (64-bit)) but only on Windows, this doesn't appear to affect Chrome on Mac OS. 

It seems the `DropZone` component now only handles drag and drop as files rather than HTML.

In our particular case we have a raw transform handler:

```
transforms: {
	from: [
		{
			type: 'raw',
			isMatch: true,
			isMultiBlock: true,
			priority: 1,
			transform: function,
		},
	],
},
```

This never gets triggered on this version of Chrome. We have found a workaround which involves adding a new transform

```
transforms: {
	from: [
		{
			type: 'files',
			isMatch: true,
			isMultiBlock: true,
			priority: 1,
			transform: function,
		},
	],
},
```

### Step-by-step reproduction instructions

1. Drag and drop into a Gutenberg DropZone component

### Screenshots, screen recording, code snippet

_No response_

### Environment info

Gutenberg 10.7.4
Windows 10 / 11
Chrome Version 96.0.4664.45 (Official Build) (64-bit)

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

No

### PR DESCRIPTION
**Pull request by sathyapulse**
## Description
The File drops intercept HTML drops, and it takes precedence with the recent Windows chrome version. It causes the plugins relying on the HTML drop to stop working. Changing the order of the checks fixes the problem — Fixes #36824.

## How has this been tested?
The steps to test has been documented in issue #36824. 

## Types of changes
Changing the order of the checks fixes the problem.

## Checklist:
- [ ] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

**Review by mcsf — COMMENTED** [review]
I don’t mind switching the order of the conditions, but it’s important to understand what has changed. If `getFilesFromDataTransfer` isn’t coming back empty, what is it returning? Are there values needing pruning, or is our function not up to spec?

**Comment by sathyapulse** [comment]
> I don’t mind switching the order of the conditions, but it’s important to understand what has changed. If `getFilesFromDataTransfer` isn’t coming back empty, what is it returning? Are there values needing pruning, or is our function not up to spec?

The function `getFilesFromDataTransfer` returns the file object and is uploaded to the media library using the blob without the fix. The HTML drops aren't captured on Windows Chrome. 

The DropZone has been rewritten without a provider recently [here](https://github.com/WordPress/gutenberg/commit/b11bc8d90a610cf21b0c8a92eef349d141d71de5#), and it changed the code significantly. I'm not sure if that's a regression from the rewrite.

**Comment by mcsf** [comment]
> The function `getFilesFromDataTransfer` returns the file object and is uploaded to the media library using the blob without the fix. The HTML drops aren't captured on Windows Chrome.

Do you mean that, _when the dragged HTML object contains an image_, newer versions of browsers are being smart about it and attaching a File data-transfer item alongside the HTML item? If so, then the fix makes sense.

**Review by ellatrix — APPROVED** [review]
Sounds good. I do vaguely remember Windows attaching an image version of copied text or html (which is a bit weird), so reordering makes sense.

File: packages/block-editor/src/components/use-on-block-drop/index.js [hunk 1]
```diff
@@ -259,10 +259,10 @@ export default function useOnBlockDrop( targetRootClientId, targetBlockIndex ) {
 		const files = getFilesFromDataTransfer( event.dataTransfer );
 		const html = event.dataTransfer.getData( 'text/html' );
 
-		if ( files.length ) {
-			_onFilesDrop( files );
-		} else if ( html ) {
+		if ( html ) {
```
Comment (ellatrix) [inline packages/block-editor/src/components/use-on-block-drop/index.js:266]: Could you just add a comment here about why the order is important and what this extra image in Windows is?

**Comment by sathyapulse** [comment]
> that, _when the dragged HTML object contains an image_, newer versions of browsers are being smart about it and attaching a File data-tra

@mcsf Yes, the newer browser version on Windows treats it as a file blob object and attaches/uploads the file to the media library and adds an image block. I did some tests to confirm the browser implementation.

On Mac Chrome 96, the file object is empty, and WP adds the HTML with an external link.
On Windows Chrome 96, the file object returns the blob, and WP uploads the image to the media library.

The file object is empty when the chrome browser is downgraded to version 95 on Windows.

CC: @ellatrix

File: packages/block-editor/src/components/use-on-block-drop/index.js (hunk 1 above)
Comment (sathyapulse) [reply packages/block-editor/src/components/use-on-block-drop/index.js:266]: @ellatrix Sure, I will add a comment. Do we need to change the order [here](https://github.com/WordPress/gutenberg/pull/30310/files#diff-5130f9f28d1f67e569b9c5498d443ab737c7e5be25e7219b7cf692c4dc3efa5dR26) as well?

Comment (ellatrix) [reply packages/block-editor/src/components/use-on-block-drop/index.js:266]: I guess that would be good if a manual test reveals a problem.

Comment (mcsf) [reply packages/block-editor/src/components/use-on-block-drop/index.js:266]: @ellatrix: should the two handlers be only defined once?

Comment (mcsf) [reply packages/block-editor/src/components/use-on-block-drop/index.js:266]: @sathyapulse: I agree with Ella. Once you have fixed the order in the other handler, and added comments to each to justify the issue (possibly linking to this PR), let us know and we can merge this.

**Review by mcsf — APPROVED** [review]

File: packages/block-editor/src/components/use-on-block-drop/index.js (hunk 1 above)
Comment (sathyapulse) [reply packages/block-editor/src/components/use-on-block-drop/index.js:266]: Thank you, @mcsf and @ellatrix.

I've added the fix in the other handler and added comments [here](https://github.com/WordPress/gutenberg/pull/37151/commits/01dc086e26358b081d3ba808e6b5fb7616c7bd21). Could you please review it?

**Pull request merged**